ros2cli/ros2cli/daemon/__init__.py

```diff
@@ -120,106 +111,37 @@ def shutdown_handler():
                 shutdown = True
             server.register_function(shutdown_handler, 'system.shutdown')
 
-            print('Serving XML-RPC on %s:%d/ros2cli/' % (addr[0], addr[1]))
+            print('Serving XML-RPC on ' + get_xmlrpc_server_url(server.server_address))
             try:
                 while not shutdown:
                     server.handle_request()
             except KeyboardInterrupt:
                 pass
-        finally:
-            server.server_close()
-
-
-def get_interfaces_ip_addresses():
-    addresses_by_interfaces = {}
-    for (kind, info_list) in netifaces.gateways().items():
-        if kind not in (netifaces.AF_INET, netifaces.AF_INET6):
-            continue
-        print('Interface kind: {}, info: {}'.format(kind, info_list))
-        addresses_by_interfaces[kind] = {}
-        for info in info_list:
-            interface_name = info[1]
-            addresses_by_interfaces[kind][interface_name] = (
-                netifaces.ifaddresses(interface_name)[kind][0]['addr']
-            )
-    print('Addresses by interfaces: {}'.format(addresses_by_interfaces))
-    return addresses_by_interfaces
-
-
-class NetworkAwareNode:
-    """A direct node, that resets itself when a network interface changes."""
-
-    def __init__(self, args):
-        self.args = args
-        # TODO(ivanpauno): A race condition is possible here, since it isn't possible to know
-        # exactly which interfaces were available at node creation.
-        self.node = DirectNode(args)
-        self.addresses_at_start = get_interfaces_ip_addresses()
-
-    def __enter__(self):
-        self.node.__enter__()
-        return self
-
-    def __getattr__(self, name):
-        attr = getattr(self.node, name)
-
-        if inspect.ismethod(attr):
-            @functools.wraps(attr)
-            def wrapper(*args, **kwargs):
-                self.reset_if_addresses_changed()
-                return getattr(self.node, name)(*args, **kwargs)
-            wrapper.__signature__ = inspect.signature(attr)
-            return wrapper
-        self.reset_if_addresses_changed()
-        return attr
-
-    def __exit__(self, exc_type, exc_value, traceback):
-        self.node.__exit__(exc_type, exc_value, traceback)
-
-    def reset_if_addresses_changed(self):
-        new_addresses = get_interfaces_ip_addresses()
-        if new_addresses != self.addresses_at_start:
-            self.addresses_at_start = new_addresses
-            self.node.destroy_node()
-            rclpy.shutdown()
-            self.node = DirectNode(self.args)
-            self.node.__enter__()
-            print('Daemon node was reset')
-
-
-def get_daemon_port():
-    base_port = 11511
-    base_port += int(os.environ.get('ROS_DOMAIN_ID', 0))
-    return base_port
+    finally:
+        server.server_close()
+
+
+def main(*, argv=None):
```

Comment:
I'm confused, is the main() function being used or are we calling `serve_forever()` directly?

Reply:
I kept this for backwards compatibility. You can still invoke the daemon separately and it'll use this function. The `ros2 daemon` CLI will daemonize the `serve_forever()` function.

Reply:
sounds good to me